server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/dialects/idms/IdmsCopybookDescriptor.java

```diff
@@ -0,0 +1,159 @@
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+
+/*
+ * copyIdmsOptions
+ *     : (RECORD copyIdmsSource versionClause? (REDEFINES cobolWord)?)
+ *     | (FILE copyIdmsSource versionClause?)
+ *     | ((MAP | MAP_CONTROL) copyIdmsSource)
+ *     | (MODULE? copyIdmsSource versionClause?)
+ *     ;
+ *
+ * copyIdmsSource
+ *     : copySource
+ *     ;
+ *
+ * copySource
+ *    : (literal | cobolWord | SUBSCHEMA_NAMES) ((OF | IN) copyLibrary)?
+ *    ;
+ *
+ * copyLibrary
+ *    : literal | cobolWord
+ *    ;
+ */
+
+/**
+ * Description of IDMS copybook
+ */
+@Data
+public class IdmsCopybookDescriptor {
+  private String name;
+  private Locality usage;
+  private Locality statement;
+  private Range levelRange;
+  private int level;
+
+  /**
+   * Factory method for copybook description object
+   * @param ctx IDMS copybook rule context
+   * @param programDocumentUri uri of file
+   * @return copybook descriptor
+   */
+  public static IdmsCopybookDescriptor from(IdmsCopyParser.CopyIdmsStatementContext ctx, String programDocumentUri) {
+    IdmsCopybookDescriptor result = new IdmsCopybookDescriptor();
+    result.name = getName(ctx);
+    Range range = DialectUtils.constructRange(ctx.copyIdmsOptions().copyIdmsSource());
+
+    if (ctx.LEVEL_NUMBER() != null && ctx.LEVEL_NUMBER().getSymbol() != null) {
+      result.levelRange = extractLevelRange(ctx.LEVEL_NUMBER().getSymbol());
+      result.level = extractLevel(ctx.LEVEL_NUMBER().getSymbol());
+    }
+
+    result.usage = Locality.builder()
+            .uri(programDocumentUri)
+            .range(range)
+            .build();
+
+    result.statement = Locality.builder()
+        .uri(programDocumentUri)
+        .range(DialectUtils.constructRange(ctx))
+        .build();
+
+    return result;
+  }
+
+  /**
+   * Factory method for copybook description object
+   * @param ctx IDMS copybook rule context
+   * @param programDocumentUri uri of file
+   * @return copybook descriptor
+   */
+  public static IdmsCopybookDescriptor from(IdmsParser.CopyIdmsStatementContext ctx, String programDocumentUri) {
+    IdmsCopybookDescriptor result = new IdmsCopybookDescriptor();
+    result.name = getName(ctx);
+    Range range = DialectUtils.constructRange(ctx.copyIdmsOptions().copyIdmsSource());
+
+    if (ctx.LEVEL_NUMBER() != null && ctx.LEVEL_NUMBER().getSymbol() != null) {
+      result.levelRange = extractLevelRange(ctx.LEVEL_NUMBER().getSymbol());
+      result.level = extractLevel(ctx.LEVEL_NUMBER().getSymbol());
+    }
+
+    result.usage = Locality.builder()
+        .uri(programDocumentUri)
+        .range(range)
+        .build();
+
+    result.statement = Locality.builder()
+        .uri(programDocumentUri)
+        .range(DialectUtils.constructRange(ctx))
+        .build();
+
+    return result;
+  }
+
+  private static int extractLevel(Token token) {
+    return Integer.parseInt(token.getText());
+  }
+
+  private static Range extractLevelRange(Token token) {
```

Comment:
Looks like a generic method. Should we shift it to some utility and let only IDMS related responsibility be handled here?

Reply:
Moved